File: python/grep/grep.py
def grep(pattern, flags, files):
    collect = []
    flags = _parse_flags(flags, files)

    for file in files:
        collect.append(_grep_on_file(pattern, flags, file))

    return "".join([y for x in collect for y in x])

def _parse_flags(flags, files):
    return {
        'ignore_case': 'i' in flags,
        'line_number': 'n' in flags,
        'file_only': 'l' in flags,
        'entire_line': 'x' in flags,
        'inverse': 'v' in flags,
        'single_file': len(files) == 1
    }


def _format(line_number, line, file, flags):
    if flags['file_only']:
        return file + '\n'
    if flags['line_number']:
        line = f"{line_number}:{line}"

    return f"{line}" if flags['single_file'] else f"{file}:{line}"


def _grep_on_file(pattern, flags, file):
    collect = []
    with open(file, 'r') as file_handle:
        for i, line in enumerate(file_handle):
            if _line_match(pattern, line, flags):
                collect.append(_format(str(i+1), line, file, flags))
                if flags['file_only']:
                    break

    return collect

def _line_match(pattern, line, flags):
    if flags['ignore_case']:
        pattern = pattern.lower()
        line = line.lower()

    if flags['entire_line']:
        pattern += '\n'
        matched = pattern == line
    else:
        matched = pattern in line

    return not matched if flags['inverse'] else matched

File: elixir/grep/lib/grep.ex
defmodule Grep do
  @option_line_number "-n"
  @option_file_name_only "-l"
  @option_case_insensitive "-i"
  @option_entire_line "-x"
  @option_inverted "-v"
  @option_add_file_name "-#"

  @spec grep(String.t(), [String.t()], [String.t()]) :: String.t()
  def grep(pattern, flags \\ [], files) do
    flags = if length(files) > 1, do: [@option_add_file_name | flags], else: flags
    flags = Enum.reduce(flags, %{}, fn x, acc -> Map.put(acc, x, true) end)
    pattern = build_pattern(pattern, flags)

    files
    |> Stream.map(fn file -> {file, grep_in_file(pattern, flags, file)} end)
    |> Stream.map(fn results -> format(results, flags) end)
    |> Enum.join("")
  end

  defp grep_in_file(pattern, flags, file) do
    filter_fn = if flags[@option_inverted], do: &Stream.reject/2, else: &Stream.filter/2

    file
    |> File.stream!()
    |> Stream.with_index(1)
    |> filter_fn.(fn {text, _line_number} -> String.match?(text, pattern) end)
    |> Enum.to_list()
  end

  defp build_pattern(pattern, flags) do
    regex_flags = if flags[@option_case_insensitive], do: "i", else: ""
    pattern = if flags[@option_entire_line], do: "^#{pattern}$", else: pattern
    Regex.compile!(pattern, regex_flags)
  end

  defp format({_file, []}, %{@option_file_name_only => true}), do: []
  defp format({file, _}, %{@option_file_name_only => true}), do: "#{file}\n"

  defp format({file, results}, flags) do
    Enum.map(results, fn {text, line_number} ->
      cond do
        flags[@option_add_file_name] && flags[@option_line_number] -> [file, line_number, text]
        flags[@option_add_file_name] -> [file, text]
        flags[@option_line_number] -> [line_number, text]
        true -> [text]
      end
      |> Enum.join(":")
    end)
  end
end
